File: 백준/Gold/14499. 주사위 굴리기/주사위 굴리기.java
import java.io.*;
import java.util.StringTokenizer;

public class Main {
    // 동, 서, 남, 북 = (1, 2, 3, 4)
    private static final int EAST = 1;
    private static final int WEST = 2;
    private static final int NORTH = 3;
    private static final int SOUTH = 4;
    private static int[] dx = {0, 0, 0, -1, 1};
    private static int[] dy = {0, 1, -1, 0, 0};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int x = Integer.parseInt(st.nextToken());
        int y = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());
        int[][] board = new int[N][M];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        /**
         *   0 북 0  0
         *  서 위 동 아래
         *   0 남 0  0
         *  0 아래 0 0
         */
        int[][] dice = new int[4][4];
        st = new StringTokenizer(br.readLine());
        while (K-- > 0) {
            int d = Integer.parseInt(st.nextToken());
            int nx = x + dx[d];
            int ny = y + dy[d];
            if (nx < 0 || nx >= N || ny < 0 || ny >= M) continue;
            if (d == EAST) {
                int tmp = dice[1][3];
                for (int i = 3; i > 0; i--) {
                    dice[1][i] = dice[1][i - 1];
                }
                dice[1][0] = tmp;
                dice[3][1] = dice[1][3];
            } else if (d == WEST) {
                int tmp = dice[1][0];
                for (int i = 0; i < 3; i++) {
                    dice[1][i] = dice[1][i + 1];
                }
                dice[1][3] = tmp;
                dice[3][1] = dice[1][3];
            } else if (d == NORTH) {
                int tmp = dice[0][1];
                for (int i = 0; i < 3; i++) {
                    dice[i][1] = dice[i + 1][1];
                }
                dice[3][1] = tmp;
                dice[1][3] = dice[3][1];
            } else if (d == SOUTH) {
                int tmp = dice[3][1];
                for (int i = 3; i > 0; i--) {
                    dice[i][1] = dice[i - 1][1];
                }
                dice[0][1] = tmp;
                dice[1][3] = dice[3][1];
            }
            if (board[nx][ny] == 0) {
                board[nx][ny] = dice[3][1];
            } else {
                dice[3][1] = dice[1][3] = board[nx][ny];
                board[nx][ny] = 0;
            }
            bw.write(dice[1][1] + "\n");
            x = nx;
            y = ny;
        }
        bw.flush();
        br.close();
        bw.close();
    }
}

File: 백준/Gold/14499. 주사위 굴리기/주사위 굴리기.py
global x, y
N, M, x, y, K = map(int, input().split())
board = [list(map(int, input().split())) for _ in range(N)]
directions = list(map(int, input().split()))

# (X, 동, 서, 북, 남)
dx = (0, 0, 0, -1, 1)
dy = (0, 1, -1, 0, 0)
dice = [[0] * 4 for _ in range(4)]

def R():
    tmp = dice[1][3]
    for i in range(3, 0, -1):
        dice[1][i] = dice[1][i - 1]
    dice[1][0] = tmp
    dice[3][1] = dice[1][3]

def L():
    tmp = dice[1][0]
    for i in range(3):
        dice[1][i] = dice[1][i + 1]
    dice[3][1] = dice[1][3] = tmp

def U():
    tmp = dice[0][1]
    for i in range(3):
        dice[i][1] = dice[i + 1][1]
    dice[1][3] = dice[3][1] = tmp

def D():
    tmp = dice[3][1]
    for i in range(3, 0, -1):
        dice[i][1] = dice[i - 1][1]
    dice[0][1] = tmp
    dice[1][3] = dice[3][1]

roll = {1:R, 2:L, 3:U, 4:D}

for d in directions:
    nx = x + dx[d]
    ny = y + dy[d]
    # 지도의 바깥으로 이동시키려고 하는 경우 명령 무시
    if nx < 0 or nx >= N or ny < 0 or ny >= M: continue
    x, y= nx, ny
    
    roll[d]()

    # 이동한 칸에 쓰여 있는 수가 0이면, 주사위의 바닥면에 쓰여 있는 수가 칸에 복사된다.
    if board[x][y] == 0:
        board[x][y] = dice[3][1]
    # 0이 아닌 경우에는 칸에 쓰여 있는 수가 주사위의 바닥면으로 복사되며, 칸에 쓰여 있는 수는 0이 된다.
    else:
        dice[3][1] = dice[1][3] = board[x][y]
        board[x][y] = 0
    print(dice[1][1])